client.go

```diff
@@ -21,15 +21,26 @@ import (
 	"github.com/tailscale/hujson"
 )
 
+// Authentication is a pluggable mechanism for authenticating requests.
+type Authentication interface {
+	// HTTPClient builds an http.Client that uses orig as a starting point and
+	// adds its own authentication to outgoing requests. baseURL is the base URL
+	// of the API server to which we will be authenticating.
+	HTTPClient(orig *http.Client, baseURL string) *http.Client
+}
+
 // Client is used to perform actions against the Tailscale API.
 type Client struct {
 	// BaseURL is the base URL for accessing the Tailscale API server. Defaults to https://api.tailscale.com.
 	BaseURL *url.URL
 	// UserAgent configures the User-Agent HTTP header for requests. Defaults to "tailscale-client-go".
 	UserAgent string
 	// APIKey allows specifying an APIKey to use for authentication.
-	// To use OAuth Client credentials, construct an [http.Client] using [OAuthConfig] and specify that below.
+	// To use OAuth Client credentials, specify OAuthCredentials instead.
 	APIKey string
+	// Authentication specifies a mechanism for adding authentication to outgoing requests.
+	// If provided, APIKey is ignored.
+	Authentication Authentication
```

Comment:
This is now pluggable. I'm using it for OAuth, we can also implement it for WIF, etc.